src/engraving/dom/systemlock.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-Studio-CLA-applies
+ *
+ * MuseScore Studio
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore Limited
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef MU_ENGRAVING_SYSTEM_LOCK_H
+#define MU_ENGRAVING_SYSTEM_LOCK_H
+
+#include <map>
+
+#include "measurebase.h"
+
+namespace mu::engraving {
+class SystemLock
+{
+public:
+    SystemLock(MeasureBase* start, MeasureBase* end)
+        : m_startMB(start), m_endMB(end)
+    {
+        assert(m_startMB->isMeasure() || m_startMB->isHBox());
+        assert(m_endMB->isMeasure() || m_endMB->isHBox());
+        assert(m_startMB->isBefore(m_endMB) || m_startMB == m_endMB);
+    }
+
+    MeasureBase* startMB() const { return m_startMB; }
+    MeasureBase* endMB() const { return m_endMB; }
+
+    bool contains(const MeasureBase* mb) const;
+
+private:
+    MeasureBase* m_startMB;
+    MeasureBase* m_endMB;
+};
+
+class SystemLocks
+{
+public:
+    void add(const SystemLock* lock);
+    void remove(const SystemLock* lock);
+    void removeLockStartingAt(const MeasureBase* mb) { m_systemLocks.erase(mb); }
+    void clear() { m_systemLocks.clear(); }
+
+    const SystemLock* lockStartingAt(const MeasureBase* mb) const;
+    const SystemLock* lockContaining(const MeasureBase* mb) const;
+    std::vector<const SystemLock*> locksContainedInRange(const MeasureBase* start, const MeasureBase* end);
+
+    std::vector<const SystemLock*> allLocks() const;
+
+private:
+#ifndef NDEBUG
+    void sanityCheck();
+    void dump();
+#endif
+
+    struct Ordering
+    {
+        bool operator()(const MeasureBase* a, const MeasureBase* b) const
+        {
+            return a->isBefore(b);
+        }
+    };
+
+    std::map<const MeasureBase*, const SystemLock*, Ordering> m_systemLocks;

```

Comment:
`MeasureBase::isBefore` seems quite expensive, so any operation on `m_systemLocks` will be quite expensive too, but this cost will be very hidden. 

For example, on a very quick first glance, it looks like `SystemLocks::lockContaining` is O(n log n) while it may be possible in O(n) by just iterating over all measures in the score. (Note: this may be nonsense, I only looked very briefly)

Reply:
> MeasureBase::isBefore seems quite expensive 

It really isn't. Because in the overwhelming majority of cases, the two measures being compared have different `m_tick`, which means isBefore will be nearly always resolved with a single tick comparison (the early return in the function). The edge case is when one or both are an HBox, in which case they have the same tick so you need to look ahead of behind, but again that's very rare, and it will usually only take just one step.
![image](https://github.com/user-attachments/assets/2e262e76-dc6a-4082-83bd-b918d30a3b30)

So no, I think this way of storing the locks is definitely more efficient :)

Reply:
Hm, I had missed the tick comparison. Then it seems fine.